README.md
# InstapaperLibrary for python
Simple library for adding items to instapaper.com

## Usage

    from instapaper_library import Instapaper
    # create instance
    i = Instapaper(login, password)
    # adding items
    i.add_item(url, title)


HISTORY.rst
History
========

0.5.1 (2016-01-13)
------------------
* fix travis.yml misconfig that only ran tests on master branch

0.5.0 (2016-01-13)
------------------
* add support for new instapaper API

0.4.0 (2010-09-06)
------------------
* add support for JSONP
* add support for redirect=close

0.3.1 (2010-06-05)
-------------------
* add doctest unit test
* catch exception in auth because of new query method

0.3.0 (2010-05-24)
-------------------
* Choose between HTTP/HTTPS (HTTPS standard)
* Set selection for added URL
* proper use of auto-title parameter
* Response headers (Content-Location, X-Instapaper-Title) are returned

0.2.0 (2010-05-20)
-------------------
* Restructured as python package
* PyPi compatibility
* Add CLI client

0.1.0 (2010-05-01)
-------------------
* Basic library working
* Working add_item


instapaper.py
##
# @file instapaper.py
# @brief simple library to use instapaper
# @author Daniel Schauenberg <d@lordofhosts.de>
# @version 0.1
# @date 2010-01-23

import urllib
import urllib2


class Instapaper:
    """ This class provides the structure for the connection object """

    def __init__(self, user, password):
        self.user = user
        self.password = password
        self.authurl = "https://www.instapaper.com/api/authenticate"
        self.addurl = "https://www.instapaper.com/api/add"

    def add_item(self, url, title=""):
        """ Method to add a new item to a instapaper account
            Returns 0 on success and -1 if something went wrong
        """
        parameters = {'username' : self.user,'password' : self.password,
                      'url' : url, 'title' : title}
        headerdata = urllib.urlencode(parameters)
        try:
            request = urllib2.Request(self.addurl, headerdata)
            response = urllib2.urlopen(request).read()
            if (int(response) == 201):
                return 0
            else:
                return -1
        except IOError:
            return -1

    def set_username(self, user):
        """ set username"""
        self.user = user

    def set_password(self, password):
        """ set password """
        self.password = password


instapaper_cli.py
#!/opt/local/bin/python2.6

from instapaper import Instapaper
from optparse import OptionParser
from getpass import getpass

def usage():
    print "Usage: instapaper.py [-h] username password url"
    print "Options:"
    print "-h   Print this help"

def main():

    # initialize parser
    usage = "usage: %prog -u USER [-t TITLE] url"
    parser = OptionParser(usage)
    parser.add_option("-u", "--user", action="store", dest="user",metavar="USER",
                      help="instapaper username")
    parser.add_option("-t", "--title", action="store", dest="title",metavar="TITLE",
                      help="title of the link to add")

    (options, args) = parser.parse_args()

    if not options.user:
        parser.error("No instapaper user given.")
    else:
        title = ""
        if options.title:
            title = options.title
        pw = getpass()
        inst = Instapaper(options.user,pw)
        result = inst.add_item(args[0],title)
        if (result == -1):
            print "Uh-Oh, something went wrong."

if __name__ == "__main__":
    main()


instapaper_library.py
##
# @file instapaper.py
# @brief simple library to use instapaper

import urllib
import urllib2

class Instapaper:
    """ This class provides the structure for the connection object """

    def __init__(self, user, password):
        self.user = user
        self.password = password
        self.authurl = "https://www.instapaper.com/api/authenticate"
        self.addurl = "https://www.instapaper.com/api/add"

    def add_item(self, url, title=""):
        """ Method to add a new item to a instapaper account
            Returns 0 on success and -1 if something went wrong
        """
        parameters = {'username' : self.user,'password' : self.password,
                      'url' : url, 'title' : title}
        headerdata = urllib.urlencode(parameters)
        try:
            request = urllib2.Request(self.addurl, headerdata)
            response = urllib2.urlopen(request).read()
            if (int(response) == 201):
                return 0
            else:
                return -1
        except IOError:
            return -1

    def set_username(self, user):
        """ set username"""
        self.user = user

    def set_password(self, password):
        """ set password """
        self.password = password


instapaperapi/instapaperapi.py
##
# @file instapaper.py
# @brief simple library to use instapaper

import urllib
import urllib2

class Instapaper:
    """ This class provides the structure for the connection object """

    def __init__(self, user, password):
        self.user = user
        self.password = password
        self.authurl = "https://www.instapaper.com/api/authenticate"
        self.addurl = "https://www.instapaper.com/api/add"

        self.add_status_codes = {
                                      201 : "URL successfully added.",
                                      400 : "Bad Request.",
                                      403 : "Invalid username or password.",
                                      500 : "Service error. Try again later."
                                }

        self.auth_status_codes = {
                                      200 : "OK.",
                                      403 : "Invalid username or password.",
                                      500 : "Service error. Try again later."
                                 }

    def add_item(self, url, title=""):
        """ Method to add a new item to a instapaper account
            Returns 0 on success and -1 if something went wrong
        """
        parameters = {'username' : self.user,'password' : self.password,
                      'url' : url, 'title' : title}
        headerdata = urllib.urlencode(parameters)
        try:
            request = urllib2.Request(self.addurl, headerdata)
            response = urllib2.urlopen(request).read()
            if (int(response) == 201):
                return 0
            else:
                return -1
        except IOError:
            return -1

    def set_username(self, user):
        """ set username"""
        self.user = user

    def set_password(self, password):
        """ set password """
        self.password = password


instapaperlib/instapaperlib.py
# encoding: utf-8
'''
 instapaperlib.py -- brief simple library to use instapaper

>>> Instapaper("instapaperlib", "").auth()
(200, 'OK.')

>>> Instapaper("instapaperlib", "dd").auth()
(200, 'OK.')

>>> Instapaper("instapaperlibi", "").auth()
(403, 'Invalid username or password.')

>>> Instapaper("instapaperlib", "").add_item("google.com")
(201, 'URL successfully added.')

>>> Instapaper("instapaperlib", "").add_item("google.com", "google")
(201, 'URL successfully added.')

>>> Instapaper("instapaperlib", "").add_item("google.com", "google", response_info=True)
(201, 'URL successfully added.', 'google', 'http://www.google.com/')

>>> Instapaper("instapaperlib", "").add_item("google.com", "google", selection="google page", response_info=True)
(201, 'URL successfully added.', 'google', 'http://www.google.com/')

>>> Instapaper("instapaperlib", "").add_item("google.com", "google", selection="google page", jsonp="callBack", response_info=True)
'callBack({"status":201,"text":"{\\\\"bookmark_id\\\\": 45744013}"})'

>>> Instapaper("instapaperlib", "").add_item("google.com", jsonp="callBack")
'callBack({"status":201,"text":"{\\\\"bookmark_id\\\\": 45744013}"})'

>>> Instapaper("instapaperlib", "").auth(jsonp="callBack")
'callBack({"status":200})'

>>> Instapaper("instapaperlib", "dd").auth(jsonp="callBack")
'callBack({"status":200})'

>>> Instapaper("instapaperlibi", "").auth(jsonp="callBack")
'callBack({"status":403})'

>>> Instapaper("instapaperlib", "").add_item("google.com", "google", redirect="close")
(201, 'URL successfully added.')

'''

import urllib
import urllib2

class Instapaper:
    """ This class provides the structure for the connection object """

    def __init__(self, user, password, https=True):
        self.user = user
        self.password = password
        if https:
            self.authurl = "https://www.instapaper.com/api/authenticate"
            self.addurl = "https://www.instapaper.com/api/add"
        else:
            self.authurl = "http://www.instapaper.com/api/authenticate"
            self.addurl = "http://www.instapaper.com/api/add"

        self.add_status_codes = {
                                      201 : "URL successfully added.",
                                      400 : "Bad Request.",
                                      403 : "Invalid username or password.",
                                      500 : "Service error. Try again later."
                                }

        self.auth_status_codes = {
                                      200 : "OK.",
                                      403 : "Invalid username or password.",
                                      500 : "Service error. Try again later."
                                 }

    def add_item(self, url, title=None, selection=None,
                 jsonp=None, redirect=None, response_info=False):
        """ Method to add a new item to a instapaper account

            Parameters: url -> URL to add
                        title -> optional title for the URL
            Returns: (status as int, status error message)
        """
        parameters = {
                      'username' : self.user,
                      'password' : self.password,
                      'url' : url,
                     }
        # look for optional parameters title and selection
        if title is not None:
            parameters['title'] = title
        else:
            parameters['auto-title'] = 1
        if selection is not None:
            parameters['selection'] = selection
        if redirect is not None:
            parameters['redirect'] = redirect
        if jsonp is not None:
            parameters['jsonp'] = jsonp

        # make query with the chosen parameters
        status, headers = self._query(self.addurl, parameters)
        # return the callback call if we want jsonp
        if jsonp is not None:
            return status
        statustxt = self.add_status_codes[int(status)]
        # if response headers are desired, return them also
        if response_info:
            return (int(status), statustxt, headers['title'], headers['location'])
        else:
            return (int(status), statustxt)

    def auth(self, user=None, password=None, jsonp=None):
        """ authenticate with the instapaper.com service

            Parameters: user -> username
                        password -> password
            Returns: (status as int, status error message)
        """
        if not user:
            user = self.user
        if not password:
            password = self.password
        parameters = {
                      'username' : self.user,
                      'password' : self.password
                     }
        if jsonp is not None:
            parameters['jsonp'] = jsonp
        status, headers = self._query(self.authurl, parameters)
        # return the callback call if we want jsonp
        if jsonp is not None:
            return status
        return (int(status), self.auth_status_codes[int(status)])

    def _query(self, url=None, params=""):
        """ method to query a URL with the given parameters

            Parameters:
                url -> URL to query
                params -> dictionary with parameter values

            Returns: HTTP response code, headers
                     If an exception occurred, headers fields are None
        """
        if url is None:
            raise NoUrlError("No URL was provided.")
        # return values
        headers = {'location': None, 'title': None}
        headerdata = urllib.urlencode(params)
        try:
            request = urllib2.Request(url, headerdata)
            response = urllib2.urlopen(request)
            # return numeric HTTP status code unless JSONP was requested
            if 'jsonp' in params:
                status = response.read()
            else:
                status = response.getcode()
            info = response.info()

            try:
                headers['location'] = info['Content-Location']
            except KeyError:
                pass
            try:
                headers['title'] = info['X-Instapaper-Title']
            except KeyError:
                pass
            return (status, headers)
        except urllib2.HTTPError as exception:
            # handle API not returning JSONP response on 403
            if 'jsonp' in params:
                return ('%s({"status":%d})' % (params['jsonp'], exception.code), headers)
            else:
                return (exception.code, headers)
        except IOError as exception:
            return (exception.code, headers)

# instapaper specific exceptions
class NoUrlError(Exception):
    """ exception to raise if no URL is given.
    """
    def __init__(self, arg):
        self.arg = arg
    def __str__(self):
        return repr(self.arg)



if __name__ == '__main__':
    import doctest
    doctest.testmod()
